File: DFS/11724/11724.cpp
#include <iostream>
#include <vector>
using namespace std;
vector<int> adj_list[1002];
bool visited[1002];
int N, M, cnt = 0;
void dfs(int u) {
    visited[u] = true;
    for (int idx = 0; idx < adj_list[u].size(); idx++) {
        int v = adj_list[u][idx];
        if (!visited[v]) {
            dfs(v);
        }
    }
}

int main() {
    cin.tie(nullptr);
    ios_base::sync_with_stdio(0);
    cin >> N >> M;
    while (M--) {
        int u, v;
        cin >> u >> v;
        adj_list[u].push_back(v);
        adj_list[v].push_back(u);
    }
    for (int idx = 1; idx <= N; idx++) {
        if (!visited[idx]) {
            dfs(idx);
            cnt++;
        }
    }
    cout << cnt << "\n";
    return 0;
}

File: DFS/11724/11724.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * DFS
 * */

private const val MAX = 1000
private val visited = BooleanArray(MAX + 1) { false }
private val adjList = Array(MAX + 1) { mutableListOf<Int>() }

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val (n, m) = readLine().split(" ").map { it.toInt() }
    repeat(m) {
        val (u, v) = readLine().split(" ").map { it.toInt() }
        adjList[u].add(v)
        adjList[v].add(u)
    }

    var cnt = 0

    for (idx in 1..n) {
        if (!visited[idx]) {
            visited[idx] = true
            dfs(idx)
            cnt++
        }
    }

    bw.write(cnt.toString())
    bw.flush()

    bw.close()
}

fun dfs(start: Int) {
    for (idx in 0 until adjList[start].size) {
        val u = adjList[start][idx]
        if (!visited[u]) {
            visited[u] = true
            dfs(u)
        }
    }
}
